src/gnn_reco/data/utils.py

```diff
@@ -9,6 +9,46 @@
 from typing import List, Tuple
 import sqlite3
 
+
+def GetDesiredEventNumbers(db_path, desired_size, fraction_noise = 0, fraction_nu_e = 0, fraction_muon = 0, fraction_nu_mu = 0, fraction_nu_tau = 0):
+    assert fraction_nu_e + fraction_muon + fraction_nu_mu + fraction_nu_tau + fraction_noise == 1.0, 'Sum of fractions not equal to one.'
+
+    fracs = [fraction_noise, fraction_nu_e, fraction_muon, fraction_nu_mu, fraction_nu_tau]
+    numbers_desired = [int(x * desired_size) for x in fracs]
+    pids = [1,12,13,14,16]
+
+    with sqlite3.connect(db_path) as con:
+        total_query = 'SELECT event_no FROM truth WHERE abs(pid) IN {}'.format(tuple(pids))
+        tot_event_nos = pd.read_sql(total_query,con)
+        if len(tot_event_nos) < desired_size:
+            desired_size = len(tot_event_nos)
+            numbers_desired = [x * desired_size for x in fracs]
+            print('Only {} events in database, using this number instead.'.format(len(tot_event_nos)))
+
+        list_of_dataframes = []
+        restart_trigger = True
+        while restart_trigger:
+            restart_trigger = False
+            for number,particle_type in zip(numbers_desired,pids):
+                query_is = 'SELECT event_no FROM truth WHERE abs(pid) == {} LIMIT {}'.format(particle_type,number)
+                dataframe = pd.read_sql(query_is,con)
+                if len(dataframe) != number:
+                    print('There have been {} requested (particle {}), we can only supply {}. \nRenormalising...'.format(number,particle_type,len(dataframe)))
+                    if len(dataframe) == 0:
+                        print('There are no particles of type {} in this database please make new request.'.format(particle_type))
+                        return None
+                    numbers_desired = [int(new_x * (len(dataframe)/number)) for new_x in numbers_desired]
+                    restart_trigger = True
+                    list_of_dataframes = []
+                    break
+                list_of_dataframes.append(dataframe)
+        
+        retrieved_event_nos_pd = pd.concat(list_of_dataframes)
+        event_no_list = retrieved_event_nos_pd.sample(frac=1, replace=False, random_state=np.random.RandomState(seed=0)).values.ravel().tolist()
```

Comment:
Here, and in conjunction with the comment regarding `LIMIT` above: We typically want these get-event-numbers-for-training functions to be random but reproducible, i.e. you should get the same (random) list of event numbers if you run this function multiple times _provided you use the same seed._ Therefore, I would propose to add an optional `seed` argument to this function, just as for the `get_equal_propoirtion_neutrino_indices` function, and use this argument in a similar manner as in that function to make the output reproducible.